crates/sui-types/src/temporary_store.rs

```diff
@@ -832,55 +683,252 @@ impl<S: ObjectStore> TemporaryStore<S> {
+        for (object, write_kind) in objects_to_update {
+            self.write_object(object, write_kind);
+        }
+        Ok(total_bytes_written_deleted as u64)
+    }
 }
 
 impl<S: GetModule + ObjectStore + BackingPackageStore> TemporaryStore<S> {
+    /// Get the total SUI in `obj` at version `v`, which should be the version before the
+    /// tx executed. for internal use only
+    fn get_input_sui(&self, id: &ObjectID, version: SequenceNumber) -> u64 {
+        if let Some(obj) = self.input_objects.get(id) {
+            assert!(obj.version() == version);
+            obj.get_total_sui(&self).unwrap()
+        } else {
+            // not in input objects, must be a dynamic field
+            let obj = self
+                .store
+                .get_object(id)
+                .unwrap()
+                .expect("Failed to look up input object and identify its total SUI");
+            assert!(obj.version() == version);
+            obj.get_total_sui(&self).unwrap()
+        }
+    }
+
     /// Check that this transaction neither creates nor destroys SUI. This should hold for all txes except
     /// the epoch change tx, which mints staking rewards equal to the gas fees burned in the previous epoch.
     /// This intended to be called *after* we have charged for gas + applied the storage rebate to the gas object,
     /// but *before* we have updated object versions
     pub fn check_sui_conserved(&self) {
-        if !self
-            .dynamic_fields_touched_do_not_call_before_fixing()
-            .is_empty()
-        {
-            // TODO: check conservation in the presence of dynamic fields
-            return;
+        let mut total_input_sui = 0;
+        let mut total_output_sui = 0;
+        for (id, (output_obj, kind)) in &self.written {
+            match kind {
+                WriteKind::Mutate => {
+                    // note: output_obj.version has not yet been increased by the tx, so output_obj.version
+                    // is the object version at tx input
+                    let input_version = output_obj.version();
+                    total_input_sui += self.get_input_sui(id, input_version);
+                    total_output_sui += output_obj.get_total_sui(&self).unwrap();
+                }
+                WriteKind::Create => {
+                    // created objects did not exist at input, and thus contribute 0 to input SUI
+                    total_output_sui += output_obj.get_total_sui(&self).unwrap();
+                }
+                WriteKind::Unwrap => {
+                    // an unwrapped object was either:
+                    // 1. wrapped in an input object A, in which case its contribution to input SUI will be captured by
+                    //    looking at A
+                    // 2. wrapped in a dynamic field, or itself a dynamic field (not sure)?
+                    // TODO: in the former case, doing nothing is the right answer, for (2) I think we need to look at the input SUI
```

Comment:
I don't think there is anything needed here for df.
If it is a df, or was in a df, the wrapper will still appear in written. Since we aren't looking at input objects, but rather through the writes, everything should be fine?

Reply:
The wrapper will appear as an output object in `written`, but the wrap-ee is now a top-level object whose SUI and storage rebate we need to account for. If we fail to account for this, a number of tests fail:

```
---- run_test::child_count/unwrap_never_stored_transfer.move ----
test panicked: assertion failed: `(left == right)`
  left: `2000019`,
 right: `2000004`: SUI conservation violated--this transaction either mints or burns SUI

---- run_test::dynamic_fields/unwrap_object.move ----
test panicked: assertion failed: `(left == right)`
  left: `2000041`,
 right: `2000025`: SUI conservation violated--this transaction either mints or burns SUI

---- run_test::sui/unwrap.move ----
test panicked: assertion failed: `(left == right)`
  left: `2000021`,
 right: `2000003`: SUI conservation violated--this transaction either mints or burns SUI

---- run_test::sui/unwrap_then_freeze.move ----
test panicked: assertion failed: `(left == right)`
  left: `2000021`,
 right: `2000003`: SUI conservation violated--this transaction either mints or burns SUI

---- run_test::transfer_object/wrap_unwrap.move ----
test panicked: assertion failed: `(left == right)`
  left: `2000019`,
 right: `2000004`: SUI conservation violated--this transaction either mints or burns SUI
```

Reply:
I was commenting more that I don't think you need the TODO, but maybe I misunderstood it